src/approaches/nsrt_rl_approach.py

```diff
@@ -49,33 +65,215 @@ def get_interaction_requests(self) -> List[InteractionRequest]:
+                curr_actions.append(next(actions))
+                curr_state_changing_feat = _create_absolute_option_param(
+                    s,
+                    parent._changing_var_to_feat,  # pylint: disable=protected-access
+                    parent._changing_var_order,  # pylint: disable=protected-access
+                    var_to_obj,
+                )
+                subgoal_state_changing_feat = curr_option.memory[
+                    "absolute_params"]
+                relative_param = subgoal_state_changing_feat - \
+                    curr_state_changing_feat
+                curr_relative_params.append(relative_param)
+
+                # In the remaining code in this loop we compute the reward.
+                reward = 0
+
+                # Add large negative reward if any object's features that are
+                # not supposed to change, do change. This can happen in two
+                # ways: 1) if any of the features of the option's objects that
+                # aren't supposed to change, change, and 2) if any feature of
+                # any other objects in the state, change.
+                prev_state_unchanging_feat = _create_absolute_option_param(
+                    curr_states[-2],
+                    var_to_unchanging_feat_ind,
+                    parent._changing_var_order,  # pylint: disable=protected-access
+                    var_to_obj,
+                )  # concatenated unchanging features of the option's objects
+                curr_state_unchanging_feat = _create_absolute_option_param(
+                    curr_states[-1],
+                    var_to_unchanging_feat_ind,
+                    parent._changing_var_order,  # pylint: disable=protected-access
+                    var_to_obj,
+                )  # concatenated unchanging features of the option's objects
+                other_objects = [
+                    o for o in s.data.keys() if o not in curr_option.objects
+                ]
+                # The first term checks the current option's objects, and the
+                # second term checks all other objects.
+                option_objects_unchanged = np.allclose(
+                    prev_state_unchanging_feat - curr_state_unchanging_feat,
+                    0,
+                    atol=1e-5)
+                other_objects_unchanged = np.allclose(
+                    curr_states[-2].vec(other_objects),
+                    curr_states[-1].vec(other_objects),
+                    atol=1e-5)
+                if not option_objects_unchanged or not other_objects_unchanged:
+                    reward += self._neg_reward * 10
+
+                # Add pos_reward if we got within epsilon of the option's
+                # subgoal, otherwise we add neg_reward.
+                # Ignore the optimization that checks for repeated states, which
+                # messes with the terminal state check.
+                _ = curr_option.memory.pop("last_state", None)
```

Comment:
modifying the memory of the option could lead to unforeseen bugs in the future. can we instead add a public method in `_LearnedNeuralParameterizedOption` that checks the abstract effects, but doesn't check the last state?

Reply:
actually, why do we even need this check? shouldn't we be rewarding the option on every time step, not just when it terminates?

Reply:
@ second question: I thought that the positive reward is only awarded for the transition into the terminal state (and then only if sufficiently close to the target subgoal). Instead, I guess I could remove the terminal state check and just check at each timestep if the current state is within epsilon of the subgoal. Then, if the effects hold only when within epsilon/2 of the subgoal, the agent may receive a positive reward multiple times in this trajectory. But that's fine– we can set epsilon however we want to to tune this. 

@ first question: any time option.terminal(s) is called, the last_state is set to s. Because we loop through the entire trajectory, we always end with last_state being set to what it originally was set to. So while we modify it, it gets modified back to the original thing properly. But I get that maybe modifying it at all might lead to bugs. The separate public method seems like a good idea, I'll do that.